File: java_pro/src/main/java/jv/com/tencent/cain/design_pattern/decorator/DecoratorPattern.java
package jv.com.tencent.cain.design_pattern.decorator;

/**
 * @author cainjiang
 * @date 2019-03-31
 */
public class DecoratorPattern {
    public static void main(String[] args) {
        Person person = new Person("程序员");
        TShirts tShirts = new TShirts();
        BigTrouser bigTrouser = new BigTrouser();
        // 不断进行装饰
        tShirts.decorate(person);
        bigTrouser.decorate(tShirts);
        bigTrouser.show();
    }
}

File: java_pro/src/main/kotlin/kt/com/tencent/cain/desigin_pattern/decorator/DecoratorPattern.kt
package kt.com.tencent.cain.desigin_pattern.decorator

/**
 * @author cainjiang
 * @date 2019-03-31
 */

fun main() {
    val person = Person("攻城狮")
    val bigTrouser = BigTrouser()
    val tShirts = TShirts()

    bigTrouser.decorate(person)
    tShirts.decorate(bigTrouser)
    tShirts.show()
}